File: sort/quick_sort.ex
defmodule QuickSort do
  # retorna um array
  @spec sort([]) :: []
  def sort(arr), do: qs(arr, 0, length(arr) - 1)

  defp qs(arr, low, high) do
    if low < high do
      {arr, pivot_idx} = partition(arr, low, high)
      qs(arr, low, pivot_idx - 1)
      qs(arr, pivot_idx + 1, high)
    else
      arr
    end
  end

  defp partition(arr, low, high) do
    pivot = arr |> Enum.at(high)

    {arr, idx} =
      Enum.reduce_while(low..(high - 1), {arr, low - 1}, fn i, {arr, idx} ->
        if arr |> Enum.at(i) <= pivot do
          new_arr = arr |> swap(idx + 1, i)
          {:cont, {new_arr, idx + 1}}
        else
          {:cont, {arr, idx}}
        end
      end)

    new_arr = arr |> swap(idx + 1, high)
    {new_arr, idx + 1}
  end

  defp swap(arr, i, j) do
    Enum.with_index(arr, fn k, val ->
      cond do
        k == i -> arr |> Enum.at(j)
        k == j -> arr |> Enum.at(i)
        true -> val
      end
    end)
  end
end

File: sort/quick-sort.kt
fun partition(arr: MutableList<Int>, low: Int, high: Int): Int {
    var pivot = arr[high]
    var idx = low - 1

    for(i in low until high) {
        if(arr[i] < pivot) {
            idx++
            val tmp = arr[i]
            arr[i] = arr[idx]
            arr[idx] = tmp
        }

    }
    idx++
    arr[high] = arr[idx]
    arr[idx] = pivot
    return idx
}

fun qs(arr: MutableList<Int>, low: Int, high: Int) {
    if(low >= high) {
        return
    }

    val pivotIdx = partition(arr, low, high)
    qs(arr, 0, pivotIdx - 1)
    qs(arr, pivotIdx + 1,high)

}


fun quickSort(arr: MutableList<Int>): List<Int> {
    qs(arr, 0, arr.size - 1)
    return arr
}

fun main() {
    val unsortedArray = mutableListOf<Int>(100, 4, 1, 5, 9, 0, 3, 34, 56, -4, 2, 201, 404, 99, 67, 26, 18)
    val sortedArray = quickSort(unsortedArray)
    println(sortedArray)
}
